docs/design/metainfo-rpc.md

```diff
@@ -0,0 +1,287 @@
+
+    rpc CreateObject(ObjectCreateRequest) returns (ObjectCreateResponse);
+    rpc CommitObject(ObjectCommitRequest) returns (ObjectCommitResponse);
+    rpc ListObjects(ObjectListRequest) returns (ObjectListResponse);
+
+    rpc DeleteObject(ObjectDeleteRequest) returns (ObjectDeleteResponse);
+    rpc FinishDeleteObject(ObjectDeleteRequest) returns (ObjectDeleteResponse);
+
+    rpc CreateSegment(SegmentCreateRequest) returns (SegmentCreateResponse);
+    rpc AbortSegment(SegmentAbortRequest) returns (SegmentAbortResponse);
+    rpc CommitSegment(SegmentCommitRequest) returns (SegmentCommitResponse);
+    rpc MakeInlineSegment(SegmentMakeInlineRequest) returns (SegmentMakeInlineResponse);
+
+    rpc Batch(BatchRequest) returns (BatchResponse);
+
+    rpc ListSegments(SegmentListRequest) returns (SegmentListResponse);
+    rpc DownloadSegment(SegmentDownloadRequest) returns (SegmentDownloadResponse);
+}
+
+message Bucket {
+    bytes        name = 1;
+    CipherSuite  path_cipher = 2;
+    bytes        attribution_id = 3;
+
+    google.protobuf.Timestamp created_at = 4;
+    
+    int64                default_segment_size = 5;
+    RedundancyScheme     default_redundancy_scheme = 6;
+    EncryptionParameters default_encryption_parameters = 7;
+}
+
+message BucketCreateRequest {
+    Bucket bucket = 1; // ignores created_at
+}
+
+message BucketCreateResponse {}
+
+message BucketGetRequest {
+    bytes name = 1;
+}
+
+message BucketGetResponse {
+    Bucket bucket = 1;
+}
+
+message BucketDeleteRequest {
+    bytes name = 1;
+}
+message BucketDeleteResponse {}
+
+message BucketListRequest {
+    bytes     cursor    = 1;
+    Direction direction = 2;
+    int32     limit     = 3;
+}
+
+message BucketListResponse {
+    repeated Bucket items = 1;
+    bool            more  = 2;
+}
+
+message Object {
+    enum Status {
+        UPLOADING;
+        COMMITTING;
+        COMMITTED;
+        DELETING;
+    }
+
+    bytes  bucket;
+    bytes  encrypted_path;
+    int32  version;
+    Status status;
+
+    bytes  stream_id;
+
+    google.protobuf.Timestamp created_at;
+    google.protobuf.Timestamp status_at;
+    google.protobuf.Timestamp expires_at;
+
+    bytes  encrypted_metadata_nonce;
+    bytes  encrypted_metadata;
+
+    int64                fixed_segment_size;
+    RedundancyScheme     redundancy_scheme;
+    EncryptionParameters encryption_parameters;
+
+    int64 total_size;
+    int64 inline_size;
+    int64 remote_size;
+}
+
+message ObjectCreateRequest {
+    bytes  bucket = 1;
+    bytes  encrypted_path = 2;
+    int32  version = 3;
+}
+
+message ObjectCreateResponse {
+    bytes  bucket;
+    bytes  encrypted_path;
+    int32  version;
+
+    bytes  stream_id;
+
+    google.protobuf.Timestamp expires_at;
+
+    bytes  encrypted_metadata_nonce;
+    bytes  encrypted_metadata;
+
+    RedundancyScheme     redundancy_scheme;
+    EncryptionParameters encryption_parameters;
```

Comment:
expires_at through encryption_parameters (with the exception of redundancy scheme) should be moved to the ObjectCreateRequest. redundancy_scheme should be moved to SegmentBeginResponse